File: src/handlers/message.js
import { logInfo } from '../utils/logging.js';
import { handleError, addReaction, sendSlackMessage } from '../utils/slack.js';
import { attachmentHandler } from '../utils/file.js';
import { handleTracebackFlag } from '../utils/slack.js';
import { invokeBedrockAgent } from '../services/bedrock.js';

// Helper function for processing messages
async function processMessage({ message, client, say, logger, text, thread = null }) {
  // Add thinking reaction
  addReaction({
    client,
    channel: message.channel,
    timestamp: message.ts,
    name: 'thinking_face',
    logger
  });

  // Check for traceback flag
  const { includeTraceback, inputText } = handleTracebackFlag(text);

  // Retrieve any file attachments
  const fileAttachments = await attachmentHandler({message, logger});

  // Get response from Bedrock with any attachments
  await sendAgentRequest({
    logger,
    say,
    client,
    channel: message.channel,
    timestamp: message.ts,
    thread,
    inputText,
    attachments: fileAttachments,
    includeTraceback
  });
}

// Helper function for sending agent requests
async function sendAgentRequest({client, channel, timestamp, inputText, say, attachments, thread, logger, includeTraceback = false}) {
  const hasAttachments = attachments && attachments.length > 0;

  // Get response from Bedrock
  const response = await invokeBedrockAgent({
    inputText: `${inputText}${hasAttachments ? ' use these files when generating your answer' : ''}`,
    attachments: hasAttachments ? attachments : null,
    sessionId: thread || timestamp,
    includeTraceback
  });

  // Check if the response is an error
  if (response?.error) {
    // Handle Bedrock agent invocation error
    addReaction({client, channel, timestamp, name: 'x', logger});
    sendSlackMessage({
      say,
      thread_ts: timestamp,
      text: `Error invoking Bedrock agent: ${response.error}`,
      logger
    });
    return;
  }

  // Post the successful response in thread
  addReaction({client, channel, timestamp, name: 'white_check_mark', logger});
  sendSlackMessage({say, thread_ts: timestamp, text: response, logger});
}

// Handle app mentions
export function handleAppMention({ event, client, say, logger }) {
  try {
    logInfo(logger, `Processing app mention: ${event.text}`);

    // Extract text without the mention
    const mentionPattern = /<@[^>]+>/;
    const mentionMatch = event.text.match(mentionPattern);
    const textAfterMention = event.text.slice(mentionMatch.index + mentionMatch[0].length).trim();

    return processMessage({
      message: event,
      client,
      say,
      logger,
      text: textAfterMention,
      thread: event.thread_ts || null
    });
  } catch (error) {
    return handleError({
      error,
      client,
      channel: event.channel,
      timestamp: event.ts,
      say,
      logger,
      messageId: event.client_msg_id
    });
  }
}

// Handle direct messages
export function handleDirectMessage({ message, client, say, logger }) {
  try {
    // Handles case if message is a direct message or a thread in a direct message
    const isDM = message.channel_type === 'im';
    const isThreadInDM = message.thread_ts && message.channel_type === 'im';

    // Skip if not a relevant message
    if ((!isDM && !isThreadInDM) ||
        message.bot_id ||
        (message.subtype && message.subtype !== 'file_share')) {
      return;
    }

    logInfo(logger, `Processing direct message: ${message.text}`);

    return processMessage({
      message,
      client,
      say,
      logger,
      text: message.text,
      thread: message.thread_ts || null
    });
  } catch (error) {
    return handleError({
      error,
      client,
      channel: message.channel,
      timestamp: message.ts,
      say,
      logger,
      messageId: message.client_msg_id
    });
  }
}

// Handle thread messages with "Hey Ragbot" in channels (not DMs)
export function handleThreadMessage({ message, client, say, logger }) {
  try {
    if (!message.thread_ts ||
      message.channel_type === 'im' ||
      !message.text ||
      !message.text.toLowerCase().trim().startsWith("hey ragbot") ||
      message.bot_id ||
      message.subtype) {
      return;
    }

    logInfo(logger, `Processing thread message: ${message.text}`);

    // Extract text after "Hey Ragbot"
    const textAfterHeyRagbot = message.text.slice(message.text.toLowerCase().indexOf("hey ragbot") + "hey ragbot".length).trim();

    return processMessage({
      message,
      client,
      say,
      logger,
      text: textAfterHeyRagbot,
      thread: message.thread_ts
    });
  } catch (error) {
    return handleError({
      error,
      client,
      channel: message.channel,
      timestamp: message.ts,
      say,
      logger,
      messageId: message.client_msg_id
    });
  }
}

File: src/handlers/message.go
package handlers

import (
	"fmt"
	"strings"

	"github.com/slack-go/slack"
	"github.com/slack-go/slack/slackevents"

	"slack-rag-server/src-go/services"
	"slack-rag-server/src-go/types"
	"slack-rag-server/src-go/utils"
)

// MessageHandler handles Slack message events
type MessageHandler struct {
	api            *slack.Client
	bedrockService *services.BedrockService
}

// NewMessageHandler creates a new MessageHandler
func NewMessageHandler(api *slack.Client, bedrockService *services.BedrockService) *MessageHandler {
	return &MessageHandler{
		api:            api,
		bedrockService: bedrockService,
	}
}

// HandleAppMention handles app mention events
func (h *MessageHandler) HandleAppMention(event *slackevents.AppMentionEvent) {
	utils.LogInfo(fmt.Sprintf("Processing app mention: %s", event.Text))

	// Extract text without the mention
	mentionPattern := "<@[^>]+>"
	textAfterMention := event.Text

	// Find mention and remove it
	if strings.Contains(event.Text, mentionPattern) {
		parts := strings.SplitN(event.Text, ">", 2)
		if len(parts) > 1 {
			textAfterMention = strings.TrimSpace(parts[1])
		}
	}

	// Process the message
	thread := event.ThreadTimeStamp
	if thread == "" {
		thread = event.TimeStamp
	}

	h.processMessage(event.Channel, event.TimeStamp, thread, textAfterMention, event.User)
}

// HandleDirectMessage handles direct messages
func (h *MessageHandler) HandleDirectMessage(event *slackevents.MessageEvent) {
	// Skip if not applicable
	if event.BotID != "" ||
	   (event.SubType != "" && event.SubType != "file_share") {
		return
	}

	utils.LogInfo(fmt.Sprintf("Processing direct message: %s", event.Text))

	// Process the message
	thread := event.ThreadTimeStamp
	if thread == "" {
		thread = event.TimeStamp
	}

	h.processMessage(event.Channel, event.TimeStamp, thread, event.Text, event.User)
}

// HandleThreadMessage handles thread messages
func (h *MessageHandler) HandleThreadMessage(event *slackevents.MessageEvent) {
	// Skip if not applicable
	if event.ThreadTimeStamp == "" ||
	   event.ChannelType == "im" ||
	   event.BotID != "" ||
	   event.SubType != "" ||
	   !strings.HasPrefix(strings.ToLower(event.Text), "hey ragbot") {
		return
	}

	utils.LogInfo(fmt.Sprintf("Processing thread message: %s", event.Text))

	// Extract text after "Hey Ragbot"
	textAfterHeyRagbot := strings.TrimSpace(
		strings.TrimPrefix(
			strings.TrimPrefix(event.Text, "Hey Ragbot"),
			"hey ragbot",
		),
	)

	// Process the message
	h.processMessage(event.Channel, event.TimeStamp, event.ThreadTimeStamp, textAfterHeyRagbot, event.User)
}

// processMessage processes a message and invokes the Bedrock agent
func (h *MessageHandler) processMessage(channel, timestamp, thread, text, user string) {
	// Add thinking reaction
	utils.AddReaction(h.api, channel, timestamp, "thinking_face")

	// Check for traceback flag
	includeTraceback, inputText := utils.HandleTracebackFlag(text)

	// Retrieve any file attachments
	var fileAttachments []types.FileAttachment

	// Convert the appropriate event type to retrieve attachments
	// This is a simplified version since we don't have direct access to files in events API
	// In a real implementation, you would need to retrieve files from the event

	// Get response from Bedrock with any attachments
	h.sendAgentRequest(channel, timestamp, thread, inputText, fileAttachments, includeTraceback)
}

// sendAgentRequest sends a request to the Bedrock agent and handles the response
func (h *MessageHandler) sendAgentRequest(channel, timestamp, thread, inputText string, attachments []types.FileAttachment, includeTraceback bool) {
	hasAttachments := len(attachments) > 0

	// Append attachment notice to input if needed
	fullInput := inputText
	if hasAttachments {
		fullInput = fullInput + " use these files when generating your answer"
	}

	// Get response from Bedrock
	response, err := h.bedrockService.InvokeBedrockAgent(fullInput, thread, attachments, includeTraceback)
	if err != nil {
		utils.LogError(err, "Error invoking Bedrock agent")
		utils.AddReaction(h.api, channel, timestamp, "x")
		utils.SendSlackMessage(h.api, channel, "Error invoking Bedrock agent: "+err.Error(), timestamp)
		return
	}

	// Check if the response is an error
	if errorResp, ok := response.(types.ErrorResponse); ok {
		utils.AddReaction(h.api, channel, timestamp, "x")
		utils.SendSlackMessage(h.api, channel, "Error invoking Bedrock agent: "+errorResp.Error, timestamp)
		return
	}

	// Handle successful response
	utils.AddReaction(h.api, channel, timestamp, "white_check_mark")

	// Format the response based on type
	var responseText string
	if agentResp, ok := response.(types.AgentResponse); ok {
		responseText = agentResp.Response
		if agentResp.Traceback != "" {
			responseText += "\n\n" + agentResp.Traceback
		}
	} else if stringResp, ok := response.(string); ok {
		responseText = stringResp
	} else {
		responseText = fmt.Sprintf("%v", response)
	}

	utils.SendSlackMessage(h.api, channel, responseText, timestamp)
}
